From 9d8fb070e782c95821c90d0ca7372fc6d7316c78 Mon Sep 17 00:00:00 2001
From: Klacke Wikstrom <cwikstro@cisco.com>
Date: Mon, 25 Jul 2016 12:46:30 +0200
Subject: [PATCH] Security flaw http://httpoxy.org/ fixed A security flaw with
 HTTP_PROXY fixed. When we now construct the cgi env variables, we just skip
 the Proxy header. Reported by dominic@varspool.com

---
 src/yaws_cgi.erl | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/src/yaws_cgi.erl b/src/yaws_cgi.erl
index 6e628f70..42a29af6 100644
--- a/src/yaws_cgi.erl
+++ b/src/yaws_cgi.erl
@@ -368,11 +368,21 @@ build_env(Arg, Scriptfilename, Pathinfo, ExtraEnv, SC) ->
             {"HTTP_IF_NONE_MATCH", H#headers.if_none_match},
             {"HTTP_IF_UNMODIFIED_SINCE", H#headers.if_unmodified_since},
             {"HTTP_COOKIE", flatten_val(make_cookie_val(H#headers.cookie))}
-           ]++lists:map(fun({http_header,_,Var,_,Val})->{tohttp(Var),Val} end,
-                        H#headers.other)
+           ]++ other_headers(H#headers.other)
           )) ++
         Extra_CGI_Vars.
 
+other_headers(Headers) ->
+    lists:zf(fun({http_header,_,Var,_,Val}) ->
+                     case tohttp(Var) of
+                         "HTTP_PROXY" ->
+                             %% See http://httpoxy.org/
+                             false;
+                         HTTP ->
+                             {true, {HTTP,Val}}
+                     end
+             end, Headers).
+
 tohttp(X) ->
     "HTTP_"++lists:map(fun tohttp_c/1, yaws:to_list(X)).
 
